File: how-to-implement-the-discrete-fourier-transform/Dft.cs
/* 
 * Discrete Fourier transform
 * by Project Nayuki, 2017. Public domain.
 * https://www.nayuki.io/page/how-to-implement-the-discrete-fourier-transform
 */

using System;
using System.Numerics;  // For Complex


public sealed class Dft {
	
	/* 
	 * Computes the discrete Fourier transform (DFT) of the given complex vector.
	 */
	public static Complex[] computeDft(Complex[] input) {
		int n = input.Length;
		Complex[] output = new Complex[n];
		for (int k = 0; k < n; k++) {  // For each output element
			Complex sum = new Complex(0, 0);
			for (int t = 0; t < n; t++) {  // For each input element
				double angle = 2 * Math.PI * t * k / n;
				sum += input[t] * Complex.Exp(new Complex(0, -angle));
			}
			output[k] = sum;
		}
		return output;
	}
	
	
	/* 
	 * Computes the discrete Fourier transform (DFT) of the given complex vector.
	 * All the array arguments must be non-null and have the same length.
	 */
	public static void computeDft(double[] inreal, double[] inimag, double[] outreal, double[] outimag) {
		int n = inreal.Length;
		for (int k = 0; k < n; k++) {  // For each output element
			double sumreal = 0;
			double sumimag = 0;
			for (int t = 0; t < n; t++) {  // For each input element
				double angle = 2 * Math.PI * t * k / n;
				sumreal +=  inreal[t] * Math.Cos(angle) + inimag[t] * Math.Sin(angle);
				sumimag += -inreal[t] * Math.Sin(angle) + inimag[t] * Math.Cos(angle);
			}
			outreal[k] = sumreal;
			outimag[k] = sumimag;
		}
	}
	
}

File: how-to-implement-the-discrete-fourier-transform/dft.rs
/* 
 * Discrete Fourier transform (Rust)
 * by Project Nayuki, 2017. Public domain.
 * https://www.nayuki.io/page/how-to-implement-the-discrete-fourier-transform
 */


fn compute_dft(inreal: &[f64], inimag: &[f64], outreal: &mut [f64], outimag: &mut [f64]) {
	let n: usize = inreal.len();
	for k in 0 .. n {  // For each output element
		let mut sumreal: f64 = 0.0;
		let mut sumimag: f64 = 0.0;
		for t in 0 .. n {  // For each input element
			let angle: f64 = 2.0 * std::f64::consts::PI
				* (t as f64) * (k as f64) / (n as f64);
			sumreal +=  inreal[t] * angle.cos() + inimag[t] * angle.sin();
			sumimag += -inreal[t] * angle.sin() + inimag[t] * angle.cos();
		}
		outreal[k] = sumreal;
		outimag[k] = sumimag;
	}
}
